### DISCUSSION TITLE
GitHub Copilot behind a Proxy with a self-signed certificate chain does still not work

### DISCUSSION THREAD

**Discussion by Fumapps** [community/community#35544] [Copilot Conversations] [answered]
Hello, I try to use GitHub Copilot for **IntelliJ** behind a ZScaler Proxy with a **self-signed** certificate chain.

Sadly, after hours of googling I found no solution/workaround which helped. The stacktrace I get in the IntelliJ-log is the following:

```
2022-10-08 20:11:54,724 [  49172]   WARN - #c.g.c.l.a.AgentGitHubService - error retrieving device code
com.github.copilot.lang.agent.rpc.JsonRpcErrorException: Error processing command signInInitiate, request id: 4, error code: -32603
	at com.github.copilot.lang.agent.vscodeRpc.DefaultJsonRpcMessageHandler.handleErrorResponse(DefaultJsonRpcMessageHandler.java:195)
	at com.github.copilot.lang.agent.vscodeRpc.DefaultJsonRpcMessageHandler.handleJsonMessage(DefaultJsonRpcMessageHandler.java:74)
	at com.github.copilot.lang.agent.vscodeRpc.VSCodeJsonRpcParser.processPendingContent(VSCodeJsonRpcParser.java:91)
	at com.github.copilot.lang.agent.vscodeRpc.VSCodeJsonRpcParser.appendLocked(VSCodeJsonRpcParser.java:71)
	at com.github.copilot.lang.agent.vscodeRpc.VSCodeJsonRpcParser.append(VSCodeJsonRpcParser.java:39)
	at com.github.copilot.lang.agent.vscodeRpc.AgentProcessHandler$1.onTextAvailable(AgentProcessHandler.java:35)
	at jdk.internal.reflect.GeneratedMethodAccessor77.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.intellij.execution.process.ProcessHandler$2.invoke(ProcessHandler.java:248)
	at jdk.proxy2/jdk.proxy2.$Proxy58.onTextAvailable(Unknown Source)
	at com.intellij.execution.process.ProcessHandler.notifyTextAvailable(ProcessHandler.java:222)
	at com.intellij.execution.process.BaseOSProcessHandler$SimpleOutputReader.onTextAvailable(BaseOSProcessHandler.java:155)
	at com.intellij.util.io.BaseOutputReader.processInput(BaseOutputReader.java:198)
	at com.intellij.util.io.BaseOutputReader.readAvailableBlocking(BaseOutputReader.java:136)
	at com.intellij.util.io.BaseDataReader.readAvailable(BaseDataReader.java:75)
	at com.intellij.util.io.BaseDataReader.doRun(BaseDataReader.java:156)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:212)
	at com.intellij.util.io.BaseDataReader.lambda$start$0(BaseDataReader.java:51)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.github.copilot.lang.agent.rpc.JsonRpcErrorException: Request signInInitiate failed with message: self signed certificate in certificate chain, request id: 4, error code: -32603
	at com.github.copilot.lang.agent.rpc.JsonRPC.parseResponse(JsonRPC.java:61)
	at com.github.copilot.lang.agent.vscodeRpc.DefaultJsonRpcMessageHandler.handleJsonMessage(DefaultJsonRpcMessageHandler.java:72)
	... 26 more
```

Finally, I am not able to activate the plugin and I find no way to use it for IntelliJ or related JetBrains products (including PyCharm, etc.).

Some information:

* I use the latest GitHub Copilot version (1.1.33.1985-nightly), tried also the last stable one
* I use IntelliJ 2022.2 Community
* I have tried different JREs as the running Java Runtimes for my IntelliJ, always importing the self-signed root cert into the "cacerts" files
* Using Settings/Server Certificates by checking the "Accept non-trusted certificates automatically" or accepting the root certs here directly didn't help
* workaround like https://github.com/djoxy-code/Github-copilot-Fix or https://stackoverflow.com/a/72564971 didn't work for me
* my proxy settings are correct, since all other plugins/features in my IntelliJ can reach the internet without problems (e.g. installing plugins from the marketplace, using gradle to download dependencies, ...)

Note: for VSCode, I do not have the problem using the win-ca extension. I want to develop with IntelliJ, **not with VSCode**.

There seem to be other people also waiting for solutions:

* https://stackoverflow.com/questions/73639787/unable-to-use-github-copilot-on-webstorm-jetbrains-when-behind-corporate-proxy
* https://github.com/orgs/community/discussions/8866 (see people mention that they are not finding any solution for JetBrains products)

Please, can you provide any solution for this and work on this important issue?

**Comment by iamcjwilliams** [comment]
@Fumapps Our copilot team is investigating this issue

**Reply by evan-dayy** [reply]
wowowwww

**Comment by iamcjwilliams** [comment]
@Fumapps The error message unable to get local issuer certificate means a proxy that uses self-signed certificates is being used. Unfortunately, [self-signed certificates is something we currently don't support](https://github.com/community/community/discussions/32285#discussioncomment-3716497).

Support for self-signed certificates is a known limitation in our current proxy support. We understand that there's a number of Copilot users that work in environments where this is a necessary requirement, that's why we have support for self-signed certificates on our radar and aim to implement the functionality in the future. We are unable to share a specific timeline for adding self-signed certificate support at this point in time.

We'd recommend that you keep an eye on our [official GitHub public product roadmap](https://github.com/github/roadmap), our [Changelog](https://github.blog/changelog/) and our [GitHub Blog](http://github.blog/) for future updates.

**Reply by yvesarrata** [reply]
Hello there, any news on that subject ?

**Reply by ayim** [reply]
Hey! Seeing this is highly upvoted, GitHub Copilot now supports self signed certificates - look for a box to check in extension settings.

**Reply by KCraig-StepStone** [reply]
Where is this setting?

**Reply by forchannot** [reply]
> Hey! Seeing this is highly upvoted, GitHub Copilot now supports self signed certificates - look for a box to check in extension settings.

where's this setting in jetbrains pycharm?

**Comment by qazt482** [comment]
Any update?

**Reply by ayim** [reply]
Hey! Seeing this is highly upvoted, GitHub Copilot now supports self signed certificates - look for a box to check in extension settings.

**Reply by ajw170** [reply]
Can you please clarify where this setting is?

**Comment by drienkop** [comment]
Even after update to [1.1.36.2164](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/259311) which seems to "Allow users to accept non-trusted certificates" and enabling the option in Tools -> Server Certificates, the plugin is still unable to connect due to: `Request Error: self signed certificate in certificate chain`

**Reply by iamcjwilliams** [reply]
The error Extension activation failed: self signed certificate in certificate chain occurs in a corporate setup as your network intercepts the SSL traffic which leaves behind a security certificate chain that Copilot is not able to trust. You can use this [marketplace extension](https://marketplace.visualstudio.com/items?itemName=linhmtran168.mac-ca-vscode) to allow Copilot to trust the CA.

**Reply by drienkop** [reply]
@iamcjwilliams this thread is about JetBrains products, not VSCode. What you posted is a VSCode plugin, not usable for JetBrains suite.

**Reply by bmuskalla** [reply]
@drienkop Can you please ensure you upgrade to the latest `1.1.37.2188` release? It contains various bug fixes around handling certificates. And you're right, checking `Allow non-trusted certificates automatically` will disable any TLS checks. I just tried this against a proxy with a self-signed cert and a the option checked and it works fine. One gotcha to be aware if: if you change any proxy settings/accepting certificates, you have to restart intellij for those changes to be picked up as IntelliJ doesn't expose ways to react to those settings.

**Reply by drienkop** [reply]
@bmuskalla Thanks for looking into this!

- I am on `1.1.37.2188` 
![image](https://user-images.githubusercontent.com/22722450/207275778-a0327a76-c14e-40a0-b33f-e88742096ae6.png)

- Ticked the checkbox
![image](https://user-images.githubusercontent.com/22722450/207275536-68e73856-349f-46e5-9011-4b0f2bc1ff97.png)
- restarted the IDE

Yet still getting:
```
2022-12-13 09:40:45,981 [  17658]   INFO - #copilot - [INFO] [default] [2022-12-13T08:40:45.981Z] [fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex
2022-12-13 09:40:46,107 [  17784]   WARN - #copilot - [ERROR] [default] [2022-12-13T08:40:46.110Z] Request Error: self signed certificate in certificate chain
2022-12-13 09:40:46,107 [  17784]   WARN - #copilot - [ERROR] [ghostText] [2022-12-13T08:40:46.111Z] Error on ghost text request FetchError: self signed certificate in certificate chain
```

Happy to provide more logs if needed.

**Reply by careylzh** [reply]
Hi there, same error with a self signed certificate:
`Sign in failed. Reason: Request signInInitiate failed with message: self signed certificate in certificate chain, request id: 3, error code: -32603`

Using some in-house proxy, self-signed cert too. 

Please advise 😫

**Reply by dpfankuchen** [reply]
Same for me. Plugin version is 1.1.38.2229

```
2022-12-21 09:58:31,462 [  31642]   INFO - #copilot - [INFO] [solutions] [2022-12-21T08:58:31.457Z] prompt: {"prefix":"// Path: conf\\...","suffix":"","isFimEnabled":false,"promptElementRanges":[{"kind":"PathMarker","start":0,"end":59},{"kind":"BeforeCursor","start":59,"end":439}]}
2022-12-21 09:58:31,466 [  31646]   INFO - #copilot - [INFO] [default] [2022-12-21T08:58:31.459Z] [fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex
2022-12-21 09:58:31,612 [  31792]   WARN - #copilot - [ERROR] [default] [2022-12-21T08:58:31.606Z] Request Error: self signed certificate in certificate chain
```

**Reply by ebner87** [reply]
I'm on version 1.1.38.2229 and still getting the error.

It's either "self signed certifcate" or "Unable to get local issuer certificate".

Any more ideas on how to solve it?

**Reply by wanyukang** [reply]
same problem in version 1.2.4.2459, please help

**Reply by TheDevAdvisor** [reply]
same problem

**Reply by mahmutoz** [reply]
same problem :/

**Reply by jtele2** [reply]
Same problem

**Reply by ghost** [reply]
Same here, i'm happy to provide more logs & info if needed.
My specific error message is (from the idea logs):

```WARN - #com.github.copilot.lang.agent.AgentGitHubService - Failed to initiate sign in
com.github.copilot.lang.agent.SignInException: Request signInInitiate failed with message: 8156758336:error:1425F175:SSL routines:ssl_choose_client_version:inappropriate fallback:../deps/openssl/openssl/ssl/statem/statem_lib.c:2016:
, request id: 3, error code: -32603
	at com.github.copilot.lang.agent.AgentGitHubService.initiateSignIn(AgentGitHubService.java:192)
	at com.github.copilot.lang.agent.AgentGitHubService.loginInteractive(AgentGitHubService.java:80)
```

When i run a curl (with my HTTP_PROXY, HTTPS_PROXY set), it works well and I get a json response with the current timestamp and an ok status.: `curl https://copilot-proxy.githubusercontent.com/_ping`

When I run an openSSL command and craft an HTTP request in there, it works well and i get the same answer (timestamp changed of course)

```openssl s_client -showcerts -servername copilot-proxy.githubusercontent.com -connect copilot-proxy.githubusercontent.com:443 -CAfile /path/to/my/pem/with/full/chain/certs.pem -issuer_checks -verify_return_error -proxy myproxy:myport```
returns ```Verify return code: 0 (ok)```

I'm using IntelliJ ```Build #IU-231.9011.34, built on May 16, 2023, Runtime version: 17.0.6+10-b829.9 aarch64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o```, with `javax.net.ssl.trustStore` and `javax.net.ssl.trustStorePassword` set to a jks containing the full chain.
I do NOT have checked the "Accept non-trusted certificates automatically" (and do not want to - we're talking business code". I do have the business licence for copilot.

Again, I'd be glad to help !

So I don't think there's any issue with the proxy itself. Seems like an openssl+proxy handling from the plugin.

Figured i'll copy that as a first class comment as well.

**Reply by CAOnees** [reply]
Https://X.Com/Home/

**Comment by rollingmoai** [comment]
Updates on this?

**Comment by siddharthgoel88** [comment, marked as answer]
I found a solution for this which works for me. I have blogged about it at  https://sidd.io/2023/01/github-copilot-self-signed-cert-issue/

Fix is as follows :

1. Export the self-signed certificate in discussion
2. Convert it into `.pem` format if not already
3. Export the path of this `.pem` cert to `NODE_EXTRA_CA_CERTS` variable
4. Restart your Intellij and it should work

**Reply by xichen-de** [reply]
> I found a solution for this which works for me. I have blogged about it at https://sidd.io/2023/01/github-copilot-self-signed-cert-issue/
> 
> Fix is as follows :
> 
> 1. Export the self-signed certificate in discussion
> 2. Convert it into `.pem` format if not already
> 3. Export the path of this `.pem` cert to `NODE_EXTRA_CA_CERTS` variable
> 4. Restart your Intellij and it should work

Thank you for your solution. I tested it on my Windows 11 laptop and it works. I did it as follows:
1. Export the certificate in Start - Search for "Manage User Certificates" and find the certificate in "Trusted Root Certification Authorities" - "Certificates". Export it as a DER encoded binary.
2. Convert it to `.pem`. I did this using the online tool: https://www.sslshopper.com/ssl-converter.html
3. Export it to `NODE_EXTRA_CA_CERTS`: in cmd `set NODE_EXTRA_CA_CERTS=C:\path\to\your\ca.pem`.

**Reply by rollingmoai** [reply]
> I did this using the online tool: [sslshopper.com/ssl-converter.html](https://www.sslshopper.com/ssl-converter.html)

Why are you relying on an online site? The `openssl` tool is available via Git Bash on Windows (or from a list of binaries from the [OpenSSL Wiki](https://wiki.openssl.org/index.php/Binaries)).

**Reply by xichen-de** [reply]
> > I did this using the online tool: [sslshopper.com/ssl-converter.html](https://www.sslshopper.com/ssl-converter.html)
> 
> Why are you relying on an online site? The `openssl` tool is available via Git Bash on Windows.

Because my company uses SVN instead of Git :( and I don't have Git on Windows.

**Reply by rollingmoai** [reply]
Alternatively, have a look at the list of binaries in the [OpenSSL Wiki](https://wiki.openssl.org/index.php/Binaries).

**Reply by drienkop** [reply]
I can confirm this solution also works for the Remote Development feature. You only need to set the NODE_EXTRA_CA_CERTS envvar on your host.

**Reply by Miqueloti** [reply]
Thanks, works fine for me now!

**Reply by oakla** [reply]
> Export the path of this .pem cert to NODE_EXTRA_CA_CERTS variable

Are you referring to the Windows Environment Variables? If so, and NODE_EXTRA_CA_CERTS doesn't already exist, should I just create it as a new system variable?

**Reply by wilsonson** [reply]
@oakla You may try to set Windows Environment Variable as NODE_EXTRA_CA_CERTS = <Path to CA Cert>. Normally it will work with nodejs application, but unfortunately this doesn't work for the copilot agent ... at least in my case.

**Reply by oakla** [reply]
> Export the certificate in Start - Search for "Manage User Certificates" and find the certificate in "Trusted Root Certification Authorities" - "Certificates". Export it as a DER encoded binary.

@xichen-de How can you tell which certificate to export? I can see about 50 of them in the Certificates folder

**Reply by xichen-de** [reply]
@oakla Hi, I think it depends on your company. For example, for me the certification is called "[company name] Root CA [version]".

**Reply by ToanNN** [reply]
Thanks, it works for me.If anyone has trouble converting certificates in DER file format in Windows.You can use the following command
`certutil.exe -encode <der file> <pem file>`

**Reply by oakla** [reply]
Okay, OPs original instructions didn't work for me at first. After some tweaks, I finally got it working. (I'm using PyCharm on Windows 11.)

The steps compared to the OP
> 1. Export the self-signed certificate in discussion
1. From Start search "Manage user certificates". Then 'Trusted Root Certification Authorities' > 'Certificates'
2. Export as DER encode binary
> 2. Convert it into .pem format if not already 
- from cmd: `certutil.exe -encode <der file> <pem file>`
> 3. Export the path of this .pem cert to NODE_EXTRA_CA_CERTS variable
- Using the windows environment variable dialogue, I created a new NODE_EXTRA_CA_CERTS **User** variable. I suspect it didn't work earlier because I had created a system variable but my PyCharm is installed as user.
> 4. Restart your Intellij and it should work
- Success!

**Reply by Nosp1** [reply]
This worked like a charm, had to export several self signed certs, but great stuff!

**Reply by elyuty** [reply]
In order to find which certificate to use you can open e.g. github.com in Chrome and check the actual certificates (by clicking a little lock icon in the address bar). For me, it appeared to be a certificate of Kaspersky antivirus!

**Reply by Adrian-Yan16** [reply]
> Okay, OPs original instructions didn't work for me at first. After some tweaks, I finally got it working. (I'm using PyCharm on Windows 11.)
> 
> The steps compared to the OP
> 
> > 1. Export the self-signed certificate in discussion
> 
> 1. From Start search "Manage user certificates". Then 'Trusted Root Certification Authorities' > 'Certificates'
> 2. Export as DER encode binary
> 
> > 2. Convert it into .pem format if not already
> 
> * from cmd: `certutil.exe -encode <der file> <pem file>`
> 
> > 3. Export the path of this .pem cert to NODE_EXTRA_CA_CERTS variable
> 
> * Using the windows environment variable dialogue, I created a new NODE_EXTRA_CA_CERTS **User** variable. I suspect it didn't work earlier because I had created a system variable but my PyCharm is installed as user.
> 
> > 4. Restart your Intellij and it should work
> 
> * Success!

Can I translate your solution to CSDN?

**Reply by forchannot** [reply]
> Okay, OPs original instructions didn't work for me at first. After some tweaks, I finally got it working. (I'm using PyCharm on Windows 11.)
> 
> The steps compared to the OP
> 
> > 1. Export the self-signed certificate in discussion
> 
> 1. From Start search "Manage user certificates". Then 'Trusted Root Certification Authorities' > 'Certificates'
> 2. Export as DER encode binary
> 
> > 2. Convert it into .pem format if not already
> 
> * from cmd: `certutil.exe -encode <der file> <pem file>`
> 
> > 3. Export the path of this .pem cert to NODE_EXTRA_CA_CERTS variable
> 
> * Using the windows environment variable dialogue, I created a new NODE_EXTRA_CA_CERTS **User** variable. I suspect it didn't work earlier because I had created a system variable but my PyCharm is installed as user.
> 
> > 4. Restart your Intellij and it should work
> 
> * Success!

it seems doesn't work for me,  I add it  to the env NODE_EXTRA_CA_CERTS both User and System , all the env seems work fine,   but it still doesm't work for me     how can i do,  the same win11 and pycharm
![1](https://user-images.githubusercontent.com/100580891/230149959-617703a9-8891-431e-a2ff-2b61794eb41d.png)

**Reply by jonatato** [reply]
god

**Reply by hoggatt** [reply]
Don't be like me: do NOT put quotes in the env variable

**Reply by kylerroloffpremier** [reply]
As of today, for me, this solution worked temporarily to resolve the issue for ZScaler and the github copilot extension on Pycharm, but eventually the same error began again. After using this solution with a ZScaler cert earlier this week, it has inexplicably broken again.

Repeating the steps to use the most up-to-date cert no longer resolves the issue.

**Reply by JoseJacin** [reply]
It worked perfectly for me!! Thank you so much for the solution. The truth is that we were going crazy, as no one could figure out the problem. In my company, we also use Zscaller and we use IntelliJ and VSCode, and just like it happens in the main message, in VSCode it works without problems but in IntelliJ we have problems with the self-signed certificate chain.

**Comment by jeremysimmons** [comment]
Still cannot find a solution for visual studio. cancelling my subscription until this is fixed.
Seriously disappointed this is not a priority to have instructions/workaround for visual studio.
Seems like you have to be using VSCode to get any love these days...

**Reply by antimeat** [reply]
any updates on this ?

**Reply by esponges** [reply]
Yep, same issue. So sad.

**Reply by oakla** [reply]
Did the above fix of adding the certificate to "NODE_EXTRA_CA_CERTS envvar on your host" not work? To be honest, I can't get it to work, but that's probably because I don't really understand the instruction

> I found a solution for this which works for me. I have blogged about it at https://sidd.io/2023/01/github-copilot-self-signed-cert-issue/
> 
> Fix is as follows :
> 
> 1. Export the self-signed certificate in discussion
> 2. Convert it into `.pem` format if not already
> 3. Export the path of this `.pem` cert to `NODE_EXTRA_CA_CERTS` variable
> 4. Restart your Intellij and it should work

**Reply by talia-ana-nguyen** [reply]
I use Win-ca extension as said in [thread](https://stackoverflow.com/questions/71367058/self-signed-certificate-in-certificate-chain-on-github-copilot). Work nicely.

**Reply by jeremysimmons** [reply]
> Did the above fix of adding the certificate to "NODE_EXTRA_CA_CERTS envvar on your host" not work?

To be honest, I didn't try the method. I am using Visual Studio, not Visual Studio Code; naturally I assumed that the plugin for Visual Studio would have a different implementation. I suppose it is possible they could use an out-of-process host for Node, and it could work...

**Reply by jeremysimmons** [reply]
> I use Win-ca extension as said in [thread](https://stackoverflow.com/questions/71367058/self-signed-certificate-in-certificate-chain-on-github-copilot). Work nicely.

Are you using Visual Studio Professional, or Visual Studio Code?

**Reply by wilsonson** [reply]
Setting NODE_EXTRA_CA_CERTS doesn't work... but using win-ca is a workaround... It seems the copilot agent ignores the NODE_EXTRA_CA_CERTS setting...

**Reply by oakla** [reply]
@wilsonson I got it working. https://github.com/orgs/community/discussions/35544#discussioncomment-5448949

You might need to create NODE_EXTRA_CA_CERTS as a **User** variable rather than a **System** variable, or vice versa

**Reply by wilsonson** [reply]
@oakla yes, in my environment, I always set NODE_EXTRA_CA_CERTS as environment variable, and it works for other applications, just not the copilot agent... don't know why... maybe need to take some time to dig deeper.

**Reply by oakla** [reply]
@wilsonson It didn't work for me when I set it as a System variable, but did work when I tried it as a User variable. Maybe you've already tried both

**Reply by jeffsdata** [reply]
@oakla, for Visual Studio 2022 with my setup it didn't work as a User Variable but it did work as a System Variable. At least it's fixed. :)

**Comment by tgrushka** [comment]
Visual Studio Code. Why isn't there a flag to ignore certificate errors in copilot extension settings? Should be global in VS Code, but that's a Microsoft issue. This is such a ubiquitous problem, it should have a very simple and readily accessible solution.

**Reply by oakla** [reply]
> it should have a very simple and readily accessible solution

Have you tried installing the win-ca extension then changing the 'inject method' to 'append' in the extension's settings?

**Reply by Sussumu** [reply]
Thanks, the win-ca solution worked for me on vs code insider!

**Comment by F21** [comment]
For intellij and other jetbrains products, just import the certificate in `.crt` format using these instructions: https://developers.cloudflare.com/cloudflare-one/connections/connect-devices/warp/user-side-certificates/install-cloudflare-cert/#jetbrains

**Reply by SangLeeGitHub** [reply]
That is not the key issue.

**Reply by muttiangelo** [reply]
Didnt worked for me, for intellij whis one worked: https://sidd.io/2023/01/github-copilot-self-signed-cert-issue/

**Comment by jeremysimmons** [comment]
My system administrator has agreed to add the urls to the bypass SSL list. Does anyone have a complete list? I know at least this:
copilot-proxy.githubusercontent.com

Has anyone tried "CoPilot for business"? https://github.blog/2023-02-14-github-copilot-for-business-is-now-available/
It states it supports "VPN proxy support"...

**Reply by SangLeeGitHub** [reply]
https://docs.github.com/en/copilot/troubleshooting-github-copilot/troubleshooting-certificate-errors-for-github-copilot

**Reply by jeremysimmons** [reply]
@hotdeveloper how does this help? I don't see any of the hostnames listed here.

**Reply by SangLeeGitHub** [reply]
The business version is no different on this issue.

**Reply by ayim** [reply]
I've submitted a request to improve the docs.
For now, this is the list:
I will delete this message when we get this in the docs.

https://github.com/login/device/code
https://github.com/login/oauth/access_token
https://api.github.com/user
https://api.github.com/copilot_internal/v2/token
https://api.github.com/copilot_internal/notification
https://default.exp-tas.com
https://copilot-telemetry.githubusercontent.com/telemetry
https://copilot-proxy.githubusercontent.com

**Reply by jeremysimmons** [reply]
what is `https://default.exp-tas.com` this?

**Reply by ayim** [reply]
Endpoint used by various VS family products to load experimental features/variables

**Reply by jeremysimmons** [reply]
> Endpoint used by various VS family products to load experimental features/variables

Thanks for the reply.

naming things is hard I guess... `feature-flags.visualstudio.com` was taken?

**Comment by SangLeeGitHub** [comment]
If you're familiar with Docker, there's a solution. There is a docker image of the famous Squid proxy that we are familiar with. `https://hub.docker.com/r/ubuntu/squid`
If you already have Docker installed on your computer, run the command below in the DOS window. (as of April 2023)
`docker run -d --name squid-container -e TZ=UTC -p 3128:3128 ubuntu/squid:5.2-22.04_beta`

And go to Jetbrains IDE such as IntelliJ, WebStorm, Pycharm, etc. and go to `settings`. (For the latest version, use the top right gear symbol)
Go to the `plugins`. Press the gear symbol at the top of the `plugins` window. Go to `HTTP Proxy settings`. The host is `localhost`, and the port is `3128` in `manual proxy settings`. 

It works well in my environment (Windows 10, Under Zscaler 4.0.0.X, WebStorm 2023.1 Build #WS-231.8109.174, built on March 28, 2023).  Enjoy.

**Reply by pablomdo** [reply]
Thanks. It worked for me.

Here's a gist with a dockerfile and squid conf file for Ubuntu users: https://gist.github.com/kryptogeist/3a9285a62744dde2b8e11973e02e0f3a

**Reply by SangLeeGitHub** [reply]
Thank you for the Dockerfile & config for Ubuntu users.

**Reply by minhnh94** [reply]
NODE_EXTRA_CA_CERTS used to work but suddenly it doesn't work anymore. Thanks for offering a working alternative to us!!

**Reply by docteurmabuse** [reply]
Thanks.

**Reply by wilsonlee47** [reply]
This method works well! Unlike tweaking VSCode or OS certificate settings, I think this approach addresses the root cause of the problem.

**Comment by ajaygautam** [comment]
I can't get this to work!

Using IntelliJ IDEA 2021.3.2 (Ultimate Edition) with Github Copilot 1.2.6.2594

Set user level variable to (I do not have permission for system level on this corporate controlled windows 11 machine)
`NODE_EXTRA_CA_CERTS=C:\dev\stuff\copilot-proxy.githubusercontent.pem;C:\dev\stuff\github.pem`

These certificates have been downloaded as der files, and converted to pem by certutil.exe

Github copilot login works fine, but when it tries to do any autocomplete, the status goes to logged out, with the following error in the log:

```log
copilot - [ERROR] [default] [2023-05-18T11:45:53.108Z] Request Error: self signed certificate in certificate chain 
copilot - [ERROR] [ghostText] [2023-05-18T11:45:53.109Z] Error on ghost text request FetchError: self signed certificate in certificate chain 
```

Anything else anyone can recommend to try?

PS: Same error when I try with Visual Studio Code - Insiders `1.79.0-insider (user setup)` and `Github Copilot Nightly` plugin.

**Comment by ghost** [comment]
Adding my own 2cts, from my reply to another comment. And hoping githubers can help us !

Same here, i'm happy to provide more logs & info if needed.
My specific error message is (from the idea logs):

```WARN - #com.github.copilot.lang.agent.AgentGitHubService - Failed to initiate sign in
com.github.copilot.lang.agent.SignInException: Request signInInitiate failed with message: 8156758336:error:1425F175:SSL routines:ssl_choose_client_version:inappropriate fallback:../deps/openssl/openssl/ssl/statem/statem_lib.c:2016:
, request id: 3, error code: -32603
	at com.github.copilot.lang.agent.AgentGitHubService.initiateSignIn(AgentGitHubService.java:192)
	at com.github.copilot.lang.agent.AgentGitHubService.loginInteractive(AgentGitHubService.java:80)
```

When i run a curl (with my HTTP_PROXY, HTTPS_PROXY set), it works well and I get a json response with the current timestamp and an ok status.: `curl https://copilot-proxy.githubusercontent.com/_ping`

When I run an openSSL command and craft an HTTP request in there, it works well and i get the same answer (timestamp changed of course)

```openssl s_client -showcerts -servername copilot-proxy.githubusercontent.com -connect copilot-proxy.githubusercontent.com:443 -CAfile /path/to/my/pem/with/full/chain/certs.pem -issuer_checks -verify_return_error -proxy myproxy:myport```
returns ```Verify return code: 0 (ok)```

I'm using IntelliJ ```Build #IU-231.9011.34, built on May 16, 2023, Runtime version: 17.0.6+10-b829.9 aarch64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o```, with `javax.net.ssl.trustStore` and `javax.net.ssl.trustStorePassword` set to a jks containing the full chain.
I do NOT have checked the "Accept non-trusted certificates automatically" (and do not want to - we're talking business code". I do have the business licence for copilot.

Again, I'd be glad to help !

So I don't think there's any issue with the proxy itself. Seems like an openssl+proxy handling from the plugin.

**Comment by neelin1** [comment]
See this conversaiton: https://stackoverflow.com/questions/71367058/self-signed-certificate-in-certificate-chain-on-github-copilot?newreg=86e63ae2f702484ba8520c8161c6d4a2
You need to install the win-ca extension and change the settings to append mode, then restart visual studio code.

**Reply by neelin1** [reply]
I was having the above issue and this solution worked for me and others

**Comment by spbisya** [comment]
Ok, I finally found a workaround.

1. Without corporate VPN - `nslookup copilot-proxy.githubusercontent.com 8.8.8.8`

> Server:		8.8.8.8
> Address:	8.8.8.8#53
> 
> Non-authoritative answer:
> Name:	copilot-proxy.githubusercontent.com
> Address: 20.74.18.95

3. Copy 20.74.18.95 and add changes to the /etc/hosts file: `nano /etc/hosts`
4. Add address and name at the end of the file like this: 20.74.18.95 copilot-proxy.githubusercontent.com
5. Restart Intellij

Working like a charm now

**Comment by melchi45** [comment]
reference this url
https://github.com/melchi45/utility/blob/main/enable-copilot-self-signed-certificate_chain.md

**Comment by minhnh94** [comment]
none of the above workarounds has worked for me so far besides the spinning up a docker squid container as proxy server.
the NODE_EXTRA_CA_CERTS used to work but doesn't anymore. maybe a jetbrains ide update or copilot plugin update rendered it useless

**Comment by nchandra-seg** [comment]
Frustrating. Anyone using Pycharm 2023.2.1 working?
I disabled certs check in pycharm settings, still doesn't work.

**Comment by d1ck50n** [comment]
i'm running IntelliJ behind Zscaler.. and it works after steps below.

1. Set HTTP Proxy
2. Disable Copilot Plugin.
3. Restart
4. Enable Copilot Plugin
5. Restart

**Comment by withoutcat** [comment]
When using VDI, I encountered a self-signed certificate issue (I find VDI environments similar to internal corporate VPNs). My development environment is Windows 11 (VDI) with IntelliJ IDEA 2024. I resolved it by following these steps:

1. Open github.com.
2. Click on the lock🔒 icon to the left of the browser address bar on github.com.
3. Verify that the connection is secure.
4. To the left of the "Close" button, click on "Show Certificate."
5. Go to the "Details" tab.
6. Click the "Export" button.
7. Open IntelliJ IDEA.
8. Navigate to "Settings."
9. Under "Tools," select "Server Certificate."
10. Click "Add."
11. Choose the certificate you exported earlier.
12. Click "OK."

Now your GitHub Copilot should work.

**Reply by PopescuMarius** [reply]
I have done the exact thing and it does not work :(
What is even more strange is that it works with Visual Studio ...

**Reply by zxie11** [reply]
same

**Comment by muttiangelo** [comment]
> When using VDI, I encountered a self-signed certificate issue (I find VDI environments similar to internal corporate VPNs). My development environment is Windows 11 (VDI) with IntelliJ IDEA 2024. I resolved it by following these steps:
> 
> 1. Open github.com.
> 2. Click on the lock🔒 icon to the left of the browser address bar on github.com.
> 3. Verify that the connection is secure.
> 4. To the left of the "Close" button, click on "Show Certificate."
> 5. Go to the "Details" tab.
> 6. Click the "Export" button.
> 7. Open IntelliJ IDEA.
> 8. Navigate to "Settings."
> 9. Under "Tools," select "Server Certificate."
> 10. Click "Add."
> 11. Choose the certificate you exported earlier.
> 12. Click "OK."
> 
> Now your GitHub Copilot should work.

yeah that worked for me besides that I had to add my certificate via terminal.

**Comment by pendar747** [comment]
I didn't manage to get the certificate as (the OP's blog post describes) to work but disabling ssl check for node using this env var worked:
﻿```
﻿export NODE_TLS_REJECT_UNAUTHORIZED='0'
﻿```

See this [stackoverflow question](https://stackoverflow.com/questions/45088006/nodejs-error-self-signed-certificate-in-certificate-chain)
